src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1198 @@
+   {
+      auto filename = wxFileName(GetFactoryPresetsPath(mEffectClassInfo), mFactoryPresets[id] + ".vstpreset");
+      return LoadPreset(filename.GetFullPath());
+   }
+   return true;
+}
+
+bool VST3Effect::LoadFactoryDefaults()
+{
+   using namespace Steinberg;
+   if(mComponentHandler == nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo { };
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kIsReadOnly)
+            continue;
+
+         if(mComponentHandler->beginEdit(parameterInfo.id) == kResultOk)
+         {
+            auto cleanup = finally([&]{
+               mComponentHandler->endEdit(parameterInfo.id);
+            });
+            mComponentHandler->performEdit(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+         }
+         mEditController->setParamNormalized(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+      }
+   }
+
+   return true;
+}
+
+namespace
+{
+   unsigned CountChannels(Steinberg::Vst::IComponent* component,
+      const Steinberg::Vst::MediaTypes mediaType, 
+      const Steinberg::Vst::BusDirection busDirection,
+      const Steinberg::Vst::BusType busType)
+   {
+      using namespace Steinberg;
+
+      unsigned channelsCount{0};
+
+      const auto busCount = component->getBusCount(mediaType, busDirection);
+      for(auto i = 0; i < busCount; ++i)
+      {
+         Vst::BusInfo busInfo;
+         if(component->getBusInfo(mediaType, busDirection, i, busInfo) == kResultOk)
+         {
+            if(busInfo.busType == busType)
+               channelsCount += busInfo.channelCount;
+         }
+      }
+      return channelsCount;
+   }
+}
+
+unsigned VST3Effect::GetAudioInCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kInput,
+      Steinberg::Vst::kMain);
+}
+
+unsigned VST3Effect::GetAudioOutCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kOutput,
+      Steinberg::Vst::kMain);
+}
+
+int VST3Effect::GetMidiInCount()
+{
+   //Dummy
+   return 0;
+}
+
+int VST3Effect::GetMidiOutCount()
+{
+   //Dummy
+   return 0;
+}
+
+void VST3Effect::SetSampleRate(double rate)
+{
+   mSetup.sampleRate = rate;
+}
+
+size_t VST3Effect::SetBlockSize(size_t maxBlockSize)
+{
+   mSetup.maxSamplesPerBlock = 
+      static_cast<Steinberg::int32>(std::min(maxBlockSize, mUserBlockSize));
+   return mSetup.maxSamplesPerBlock;
+}
+
+size_t VST3Effect::GetBlockSize() const
+{
+   return mSetup.maxSamplesPerBlock;
+}
+
+sampleCount VST3Effect::GetLatency()
+{
+   if(mUseLatency)
+   {
+      if(!mRealtimeGroupProcessors.empty())
+         return mRealtimeGroupProcessors[0]->GetLatency();
+      auto delay = mInitialDelay;
+      mInitialDelay = 0u;
```

Comment:
This function will return zero the second time it is called.  That is strange.  Why?